// tdesign-mobile-vue/esm/tabs/tab-panel.js
/**
 * tdesign v0.12.1
 * (c) 2022 TDesign Group
 * @license MIT
 */

import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { defineComponent, inject, computed, getCurrentInstance, toRefs, resolveComponent, withDirectives, openBlock, createElementBlock, normalizeClass, renderSlot, createVNode, vShow, createCommentVNode } from 'vue';
import '../shared/index.js';
import config from '../config.js';
import TabPanelProps from './tab-panel-props.js';
import TNodeComponent from '../shared/render-tnode.js';
import { renderContent } from '../shared/render.js';
import '../shared/functions.js';
import '../shared/util.js';
import '../shared/component.js';
import '../shared/constants.js';
import '../shared/useToggle/index.js';
import '../shared/useCountDown/index.js';
import '@babel/runtime/helpers/asyncToGenerator';
import '@babel/runtime/regenerator';
import '@vueuse/core';
import '../shared/useCountDown/utils.js';
import '../shared/useDefault/index.js';
import 'lodash/camelCase';
import '../shared/useEmitEvent/index.js';
import '../shared/useChildSlots/index.js';
import '@babel/runtime/helpers/toConsumableArray';
import '../shared/useVModel/index.js';
import '../shared/useTouch/index.js';
import '../shared/useScrollParent/index.js';
import '../shared/useExpose/index.js';
import '@babel/runtime/helpers/typeof';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var prefix = config.prefix;
var name = "".concat(prefix, "-tab-panel");
var script = defineComponent({
  name: name,
  components: {
    TNode: TNodeComponent
  },
  props: TabPanelProps,
  setup: function setup(props) {
    var currentValue = inject("currentValue", "");
    var isActive = computed(function () {
      return currentValue.value === props.value;
    });
    var internalInstance = getCurrentInstance();
    var panelContent = computed(function () {
      return renderContent(internalInstance, "default", "panel");
    });
    var tabPanelClass = computed(function () {
      return ["".concat(name), "".concat(prefix, "-tabs__panel")];
    });
    return _objectSpread({
      prefix: prefix,
      name: name,
      isActive: isActive,
      panelContent: panelContent,
      tabPanelClass: tabPanelClass
    }, toRefs(props));
  }
});

var _hoisted_1 = ["value"];
function render(_ctx, _cache, $props, $setup, $data, $options) {
  var _component_t_node = resolveComponent("t-node");
  return _ctx.destroyOnHide || _ctx.isActive ? withDirectives((openBlock(), createElementBlock("div", {
    key: 0,
    class: normalizeClass(_ctx.tabPanelClass),
    value: _ctx.value
  }, [renderSlot(_ctx.$slots, "default", {}, function () {
    return [createVNode(_component_t_node, {
      content: _ctx.panelContent
    }, null, 8, ["content"])];
  })], 10, _hoisted_1)), [[vShow, _ctx.isActive]]) : createCommentVNode("", true);
}

script.render = render;

export { script as default };
