// tdesign-mobile-vue/esm/loading/props.js
/**
 * tdesign v0.12.1
 * (c) 2022 TDesign Group
 * @license MIT
 */

var LoadingProps = {
  default: {
    type: [String, Function]
  },
  content: {
    type: [String, Function]
  },
  delay: {
    type: Number,
    default: 0
  },
  duration: {
    type: Number,
    default: 800
  },
  indicator: {
    type: Boolean,
    default: true
  },
  inheritColor: Boolean,
  layout: {
    type: String,
    default: "horizontal",
    validator: function validator(val) {
      if (!val) return true;
      return ["horizontal", "vertical"].includes(val);
    }
  },
  loading: {
    type: Boolean,
    default: true
  },
  pause: Boolean,
  progress: {
    type: Number
  },
  reverse: Boolean,
  size: {
    type: String,
    default: "40rpx"
  },
  text: {
    type: String
  },
  theme: {
    type: String,
    default: "circular",
    validator: function validator(val) {
      if (!val) return true;
      return ["circular", "spinner", "bar", "error", "dots"].includes(val);
    }
  }
};

export { LoadingProps as default };
