// tdesign-mobile-vue/lib/message/index.js
/**
 * tdesign v0.12.1
 * (c) 2022 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var vue = require('vue');
var message_message = require('./message.js');
require('@babel/runtime/helpers/slicedToArray');
require('tdesign-icons-vue-next');
require('./props.js');
require('../config.js');
require('../shared/render-tnode.js');
require('../shared/useEmitEvent/index.js');
require('lodash/camelCase');
require('../shared/render.js');
require('@babel/runtime/helpers/typeof');
require('../shared/useVModel/index.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty__default["default"](target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function create(props) {
  var visible = vue.ref(false);
  var root = document.createElement("div");
  document.body.appendChild(root);
  var component = vue.defineComponent({
    render: function render() {
      return vue.h(message_message["default"], _objectSpread(_objectSpread({}, props), {}, {
        visible: visible.value,
        onClose: function onClose() {
          visible.value = false;
        },
        onClosed: function onClosed() {
          root.remove();
        }
      }));
    }
  });
  vue.createApp(component).mount(root);
  vue.nextTick(function () {
    visible.value = true;
  });
}
var defaultProps = {
  align: "left",
  closeBtn: "",
  content: "",
  duration: 3e3,
  theme: "info",
  visible: false,
  zIndex: 5e3,
  onClose: function onClose() {},
  onClosed: function onClosed() {},
  onOpen: function onOpen() {},
  onOpened: function onOpened() {},
  onChange: function onChange() {}
};
["info", "success", "warning", "error"].forEach(function (theme) {
  message_message["default"][theme] = function (options) {
    var props = _objectSpread(_objectSpread({}, defaultProps), {}, {
      theme: theme
    });
    if (typeof options === "string") {
      props.content = options;
    } else {
      props = _objectSpread(_objectSpread({}, props), options);
    }
    create(props);
  };
});
message_message["default"].install = function (app) {
  var name = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : "";
  app.component(name || message_message["default"].name, message_message["default"]);
  app.config.globalProperties.$message = message_message["default"];
};
var MessagePlugin = message_message["default"];

exports.MessagePlugin = MessagePlugin;
exports["default"] = MessagePlugin;
