// tdesign-mobile-vue/lib/message/props.js
/**
 * tdesign v0.12.1
 * (c) 2022 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var messageProps = {
  align: {
    type: String,
    default: "left",
    validator: function validator(val) {
      if (!val) return true;
      return ["left", "center"].includes(val);
    }
  },
  closeBtn: {
    type: [String, Boolean, Function],
    default: void 0
  },
  content: {
    type: [String, Function]
  },
  duration: {
    type: Number,
    default: 3e3
  },
  marquee: {
    type: [Boolean, Object],
    default: false
  },
  theme: {
    type: String,
    default: "info",
    validator: function validator(val) {
      if (!val) return true;
      return ["info", "success", "warning", "error"].includes(val);
    }
  },
  visible: {
    type: Boolean,
    default: void 0
  },
  modelValue: {
    type: Boolean,
    default: void 0
  },
  defaultVisible: Boolean,
  zIndex: {
    type: Number
  },
  onChange: Function,
  onClose: Function,
  onClosed: Function,
  onOpen: Function,
  onOpened: Function
};

exports["default"] = messageProps;
