// tdesign-mobile-vue/esm/checkbox/checkbox.js
/**
 * tdesign v0.12.1
 * (c) 2022 TDesign Group
 * @license MIT
 */

import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { defineComponent, h, getCurrentInstance, inject, computed, toRefs, resolveComponent, openBlock, createElementBlock, normalizeClass, createElementVNode, createBlock, createCommentVNode, normalizeStyle, createVNode } from 'vue';
import { MinusCircleFilledIcon, CheckCircleFilledIcon, CircleIcon } from 'tdesign-icons-vue-next';
import config from '../config.js';
import CheckboxProps from './props.js';
import '../shared/index.js';
import ClASSNAMES from '../shared/constants.js';
import TNodeComponent from '../shared/render-tnode.js';
import { useDefault } from '../shared/useDefault/index.js';
import { renderContent, renderTNode } from '../shared/render.js';
import '../shared/functions.js';
import '../shared/util.js';
import '../shared/component.js';
import '../shared/useToggle/index.js';
import '../shared/useCountDown/index.js';
import '@babel/runtime/helpers/asyncToGenerator';
import '@babel/runtime/regenerator';
import '@vueuse/core';
import '../shared/useCountDown/utils.js';
import '../shared/useEmitEvent/index.js';
import 'lodash/camelCase';
import '../shared/useChildSlots/index.js';
import '@babel/runtime/helpers/toConsumableArray';
import '../shared/useVModel/index.js';
import '../shared/useTouch/index.js';
import '../shared/useScrollParent/index.js';
import '../shared/useExpose/index.js';
import '@babel/runtime/helpers/typeof';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var prefix = config.prefix;
var name = "".concat(prefix, "-checkbox");
var script = defineComponent({
  name: name,
  components: {
    TNode: TNodeComponent,
    MinusCircleFilledIcon: MinusCircleFilledIcon
  },
  props: _objectSpread(_objectSpread({}, CheckboxProps), {}, {
    borderless: {
      type: Boolean,
      value: false
    }
  }),
  emits: ["update:checked", "update:modelValue", "change"],
  setup: function setup(props, context) {
    var flagName = name;
    var checkIcons = props.icon || [h(CheckCircleFilledIcon), h(CircleIcon)];
    var _useDefault = useDefault(props, context.emit, "checked", "change"),
      _useDefault2 = _slicedToArray(_useDefault, 2),
      innerChecked = _useDefault2[0],
      setInnerChecked = _useDefault2[1];
    var internalInstance = getCurrentInstance();
    var checkboxGroup = inject("checkboxGroup", void 0);
    var labelContent = computed(function () {
      return renderContent(internalInstance, "label", "default");
    });
    var checkboxContent = computed(function () {
      return renderTNode(internalInstance, "content");
    });
    var indeterminate = computed(function () {
      if (props.checkAll && checkboxGroup != null) return checkboxGroup.checkAllStatus.value === "indeterminate";
      return props.indeterminate;
    });
    var isChecked = computed(function () {
      if (props.checkAll) return (checkboxGroup === null || checkboxGroup === void 0 ? void 0 : checkboxGroup.checkAllStatus.value) === "checked";
      if (checkboxGroup != null && props.value != null) {
        var _checkboxGroup$checke;
        return !!((_checkboxGroup$checke = checkboxGroup.checkedSet.value) !== null && _checkboxGroup$checke !== void 0 && _checkboxGroup$checke.has(props.value));
      }
      return innerChecked.value;
    });
    var isDisabled = computed(function () {
      if (checkboxGroup !== null && checkboxGroup !== void 0 && checkboxGroup.max.value) return checkboxGroup.max.value <= checkboxGroup.innerValue.value.length && !isChecked.value;
      if (props.disabled != null) return props.disabled;
      return !!(checkboxGroup !== null && checkboxGroup !== void 0 && checkboxGroup.disabled.value);
    });
    var componentClass = computed(function () {
      var _ref;
      return ["".concat(flagName), (_ref = {}, _defineProperty(_ref, ClASSNAMES.STATUS.checked, isChecked.value), _defineProperty(_ref, ClASSNAMES.STATUS.disabled, isDisabled.value), _defineProperty(_ref, ClASSNAMES.STATUS.indeterminate, indeterminate.value), _ref)];
    });
    var getLimitRowStyle = function getLimitRowStyle(row) {
      return {
        display: "-webkit-box",
        overflow: "hidden",
        WebkitBoxOrient: "vertical",
        WebkitLineClamp: row
      };
    };
    var labelStyle = computed(function () {
      return _objectSpread({
        color: isDisabled.value ? "#dcdcdc" : "inherit"
      }, getLimitRowStyle(props.maxLabelRow));
    });
    var contentStyle = computed(function () {
      return _objectSpread({}, getLimitRowStyle(props.maxContentRow));
    });
    var handleChange = function handleChange(e, source) {
      if (isDisabled.value) return;
      if (source === "content" && props.contentDisabled) return;
      var value = !isChecked.value;
      setInnerChecked(value, {
        e: e
      });
      e.stopPropagation();
      if (checkboxGroup && checkboxGroup !== null && checkboxGroup !== void 0 && checkboxGroup.onCheckedChange) {
        checkboxGroup.onCheckedChange({
          checked: value,
          checkAll: props.checkAll,
          e: e,
          option: props
        });
      }
    };
    return _objectSpread(_objectSpread({}, toRefs(props)), {}, {
      isChecked: isChecked,
      checkIcons: checkIcons,
      labelContent: labelContent,
      labelStyle: labelStyle,
      checkboxContent: checkboxContent,
      contentStyle: contentStyle,
      isDisabled: isDisabled,
      flagName: flagName,
      componentClass: componentClass,
      indeterminate: indeterminate,
      handleChange: handleChange
    });
  }
});

var _hoisted_1 = ["name", "value", "disabled", "readonly", "checked", "indeterminate"];
var _hoisted_2 = ["name", "value", "disabled", "readonly", "checked", "indeterminate"];
function render(_ctx, _cache, $props, $setup, $data, $options) {
  var _normalizeClass2;
  var _component_t_node = resolveComponent("t-node");
  var _component_minus_circle_filled_icon = resolveComponent("minus-circle-filled-icon");
  return openBlock(), createElementBlock("div", {
    class: normalizeClass(_ctx.componentClass)
  }, [createElementVNode("div", {
    class: normalizeClass("".concat(_ctx.flagName, "__content-wrap"))
  }, [_ctx.align === "left" ? (openBlock(), createElementBlock("span", {
    key: 0,
    class: normalizeClass("".concat(_ctx.flagName, "__icon-left"))
  }, [createElementVNode("input", {
    type: "checkbox",
    name: _ctx.name,
    class: normalizeClass("".concat(_ctx.flagName, "__original-left")),
    value: _ctx.value,
    disabled: _ctx.isDisabled,
    readonly: _ctx.readonly,
    checked: _ctx.isChecked,
    indeterminate: _ctx.indeterminate,
    onClick: _cache[0] || (_cache[0] = function () {
      return _ctx.handleChange && _ctx.handleChange.apply(_ctx, arguments);
    })
  }, null, 10, _hoisted_1), !_ctx.indeterminate ? (openBlock(), createBlock(_component_t_node, {
    key: 0,
    content: _ctx.checkIcons[_ctx.isChecked ? 0 : 1]
  }, null, 8, ["content"])) : (openBlock(), createBlock(_component_minus_circle_filled_icon, {
    key: 1
  }))], 2)) : createCommentVNode("", true), _ctx.labelContent || _ctx.checkboxContent ? (openBlock(), createElementBlock("span", {
    key: 1,
    class: normalizeClass((_normalizeClass2 = {}, _defineProperty(_normalizeClass2, "".concat(_ctx.flagName, "__label"), true), _defineProperty(_normalizeClass2, "".concat(_ctx.flagName, "__label-left"), _ctx.align === "right"), _normalizeClass2)),
    onClick: _cache[1] || (_cache[1] = function (e) {
      return _ctx.handleChange(e, "content");
    })
  }, [_ctx.labelContent ? (openBlock(), createElementBlock("span", {
    key: 0,
    style: normalizeStyle(_ctx.labelStyle)
  }, [createVNode(_component_t_node, {
    content: _ctx.labelContent
  }, null, 8, ["content"])], 4)) : createCommentVNode("", true), _ctx.checkboxContent ? (openBlock(), createElementBlock("span", {
    key: 1,
    class: normalizeClass("".concat(_ctx.flagName, "__description")),
    style: normalizeStyle(_ctx.contentStyle)
  }, [createVNode(_component_t_node, {
    content: _ctx.checkboxContent
  }, null, 8, ["content"])], 6)) : createCommentVNode("", true)], 2)) : createCommentVNode("", true), _ctx.align === "right" ? (openBlock(), createElementBlock("span", {
    key: 2,
    class: normalizeClass("".concat(_ctx.flagName, "__icon-right"))
  }, [createElementVNode("input", {
    type: "checkbox",
    name: _ctx.name,
    class: normalizeClass("".concat(_ctx.flagName, "__original-right")),
    value: _ctx.value,
    disabled: _ctx.isDisabled,
    readonly: _ctx.readonly,
    checked: _ctx.isChecked,
    indeterminate: _ctx.indeterminate,
    onClick: _cache[2] || (_cache[2] = function () {
      return _ctx.handleChange && _ctx.handleChange.apply(_ctx, arguments);
    })
  }, null, 10, _hoisted_2), !_ctx.indeterminate ? (openBlock(), createBlock(_component_t_node, {
    key: 0,
    content: _ctx.checkIcons[_ctx.isChecked ? 0 : 1]
  }, null, 8, ["content"])) : (openBlock(), createBlock(_component_minus_circle_filled_icon, {
    key: 1
  }))], 2)) : createCommentVNode("", true)], 2), !_ctx.borderless ? (openBlock(), createElementBlock("div", {
    key: 0,
    class: normalizeClass("".concat(_ctx.flagName, "__border ").concat(_ctx.flagName, "__border--").concat(_ctx.align))
  }, null, 2)) : createCommentVNode("", true)], 2);
}

script.render = render;

export { script as default };
