// tdesign-mobile-vue/esm/checkbox/props.js
/**
 * tdesign v0.12.1
 * (c) 2022 TDesign Group
 * @license MIT
 */

var CheckboxProps = {
  align: {
    type: String,
    default: "left",
    validator: function validator(val) {
      if (!val) return true;
      return ["left", "right"].includes(val);
    }
  },
  checkAll: Boolean,
  checked: {
    type: Boolean,
    default: void 0
  },
  modelValue: {
    type: Boolean,
    default: void 0
  },
  defaultChecked: Boolean,
  content: {
    type: [String, Function]
  },
  contentDisabled: Boolean,
  default: {
    type: [String, Function]
  },
  disabled: {
    type: Boolean,
    default: void 0
  },
  icon: {
    type: Array
  },
  indeterminate: Boolean,
  label: {
    type: [String, Function]
  },
  maxContentRow: {
    type: Number,
    default: 5
  },
  maxLabelRow: {
    type: Number,
    default: 3
  },
  name: {
    type: String,
    default: ""
  },
  readonly: Boolean,
  value: {
    type: [String, Number]
  },
  onChange: Function
};

export { CheckboxProps as default };
