// tdesign-mobile-vue/lib/_chunks/dep-5f813a8c.js
/**
 * tdesign v0.12.1
 * (c) 2022 TDesign Group
 * @license MIT
 */

'use strict';

var vue = require('vue');

var _hoisted_1 = {
  key: 0,
  class: "t-cascader__steps"
};
var _hoisted_2 = ["onClick"];
var script = /* @__PURE__ */vue.defineComponent({
  __name: "steps",
  props: {
    items: Array,
    value: Number
  },
  emits: ["update:value"],
  setup: function setup(__props, _ref) {
    var emit = _ref.emit;
    var handleClick = function handleClick(index) {
      emit("update:value", index);
    };
    return function (_ctx, _cache) {
      return Array.isArray(__props.items) ? (vue.openBlock(), vue.createElementBlock("div", _hoisted_1, [(vue.openBlock(true), vue.createElementBlock(vue.Fragment, null, vue.renderList(__props.items, function (item, index) {
        return vue.openBlock(), vue.createElementBlock("div", {
          key: index,
          class: "t-cascader__step",
          onClick: function onClick($event) {
            return handleClick(index);
          }
        }, [vue.createElementVNode("div", {
          class: vue.normalizeClass({
            "t-cascader__step-dot": true,
            "t-cascader__step-dot--active": index === __props.value,
            "t-cascader__step-dot--last": index === __props.items.length - 1
          })
        }, null, 2), vue.createElementVNode("div", {
          class: vue.normalizeClass({
            "t-cascader__step-label": true,
            "t-cascader__step-label--active": index === __props.value
          })
        }, vue.toDisplayString(item), 3)], 8, _hoisted_2);
      }), 128))])) : vue.createCommentVNode("", true);
    };
  }
});

exports.script = script;
