// tdesign-mobile-vue/es/tag/props.js
/**
 * tdesign v0.12.1
 * (c) 2022 TDesign Group
 * @license MIT
 */

var TagProps = {
  closable: Boolean,
  content: {
    type: [String, Number, Function]
  },
  disabled: Boolean,
  icon: {
    type: Function,
    default: void 0
  },
  maxWidth: {
    type: [String, Number]
  },
  shape: {
    type: String,
    default: "square",
    validator: function validator(val) {
      return ["square", "round", "mark"].includes(val);
    }
  },
  size: {
    type: String,
    default: "medium",
    validator: function validator(val) {
      return ["small", "medium", "large"].includes(val);
    }
  },
  theme: {
    type: String,
    default: "default",
    validator: function validator(val) {
      return ["default", "primary", "warning", "danger", "success"].includes(val);
    }
  },
  variant: {
    type: String,
    default: "dark",
    validator: function validator(val) {
      return ["dark", "light", "outline", "light-outline"].includes(val);
    }
  },
  onClick: Function,
  onClose: Function
};

export { TagProps as default };
