// tdesign-mobile-vue/lib/grid/grid.js
/**
 * tdesign v0.12.1
 * (c) 2022 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var vue = require('vue');
var config = require('../config.js');
var grid_props = require('./props.js');

var prefix = config["default"].prefix;
var name = "".concat(prefix, "-grid");
var script = vue.defineComponent({
  name: name,
  props: grid_props["default"],
  setup: function setup(props) {
    var _toRefs = vue.toRefs(props),
      column = _toRefs.column,
      gutter = _toRefs.gutter,
      border = _toRefs.border,
      align = _toRefs.align;
    vue.provide("grid", {
      column: column,
      gutter: gutter,
      border: border,
      align: align
    });
    return {
      name: name
    };
  }
});

function render(_ctx, _cache, $props, $setup, $data, $options) {
  return vue.openBlock(), vue.createElementBlock("div", {
    class: vue.normalizeClass("".concat(_ctx.name))
  }, [vue.renderSlot(_ctx.$slots, "default")], 2);
}

script.render = render;

exports["default"] = script;
