// tdesign-mobile-vue/esm/action-sheet/menu-grid.js
/**
 * tdesign v0.12.1
 * (c) 2022 TDesign Group
 * @license MIT
 */

import { defineComponent, ref, computed, resolveComponent, openBlock, createElementBlock, normalizeClass, createElementVNode, normalizeStyle, Fragment, renderList, createVNode, withCtx, createBlock, createCommentVNode } from 'vue';
import config from '../config.js';
import { Grid, GridItem } from '../grid/index.js';
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import '../grid/grid.js';
import '../grid/props.js';
import '../grid/grid-item.js';
import '../badge/index.js';
import '../badge/badge.js';
import '@babel/runtime/helpers/slicedToArray';
import '../shared/index.js';
import '../shared/functions.js';
import '../shared/util.js';
import '../shared/component.js';
import '../shared/constants.js';
import '../shared/render.js';
import '@babel/runtime/helpers/typeof';
import 'lodash/camelCase';
import '../shared/render-tnode.js';
import '../shared/useToggle/index.js';
import '../shared/useCountDown/index.js';
import '@babel/runtime/helpers/asyncToGenerator';
import '@babel/runtime/regenerator';
import '@vueuse/core';
import '../shared/useCountDown/utils.js';
import '../shared/useDefault/index.js';
import '../shared/useEmitEvent/index.js';
import '../shared/useChildSlots/index.js';
import '@babel/runtime/helpers/toConsumableArray';
import '../shared/useVModel/index.js';
import '../shared/useTouch/index.js';
import '../shared/useScrollParent/index.js';
import '../shared/useExpose/index.js';
import '../badge/props.js';
import '../badge/style';
import '../badge/type.js';
import '../grid/grid-item-props.js';
import '../grid/style';
import '../grid/type.js';

var prefix = config.prefix;
var name = "".concat(prefix, "-action-sheet");
var script = defineComponent({
  components: {
    TGrid: Grid,
    TGridItem: GridItem
  },
  props: {
    items: {
      type: Array,
      required: true
    },
    count: {
      type: Number,
      default: 8
    }
  },
  emits: ["selected"],
  setup: function setup(props, context) {
    var containerWrapper = ref(null);
    var moveOffset = ref(0);
    var currentIndex = ref(0);
    var useTransition = ref(true);
    var startX = 0;
    var startOffset = 0;
    var canMove = true;
    var wrapperStyle = computed(function () {
      return {
        transform: "translate3d(".concat(moveOffset.value, "px, 0, 0)"),
        transition: useTransition.value ? "transform 300ms" : "all"
      };
    });
    var gridColumn = computed(function () {
      return Math.ceil(props.count / 2);
    });
    var pageNum = computed(function () {
      return Math.ceil(props.items.length / props.count);
    });
    var actionItems = computed(function () {
      var res = [];
      for (var i = 0; i < pageNum.value; i++) {
        var temp = props.items.slice(i * props.count, (i + 1) * props.count);
        res.push(temp);
      }
      return res;
    });
    var handleSelected = function handleSelected(index) {
      context.emit("selected", index);
    };
    var moveByIndex = function moveByIndex(index) {
      useTransition.value = true;
      if (containerWrapper.value) {
        moveOffset.value = pageNum.value > 1 ? index * containerWrapper.value.offsetWidth * -1 : 0;
      }
    };
    var getMaxOffset = function getMaxOffset() {
      if (!containerWrapper.value) return 0;
      return (pageNum.value - 1) * containerWrapper.value.offsetWidth;
    };
    var handleTouchstart = function handleTouchstart(e) {
      canMove = true;
      useTransition.value = false;
      startX = e.touches[0].clientX;
      startOffset = startX - moveOffset.value;
    };
    var handleTouchmove = function handleTouchmove(e) {
      var clientX = e.touches[0].clientX;
      var minOffset = 0;
      var maxOffset = getMaxOffset();
      if (Math.abs(clientX - startX) < 15) return;
      moveOffset.value = clientX - startOffset;
      if (moveOffset.value > minOffset) {
        moveOffset.value = minOffset;
        canMove = false;
      }
      if (Math.abs(moveOffset.value) >= maxOffset) {
        moveOffset.value = maxOffset * -1;
        canMove = false;
      }
    };
    var handleTouchend = function handleTouchend(e) {
      if (!canMove) return;
      var distance = e.changedTouches[0].clientX - startX;
      var targetIndex = Math.abs(distance) > 50 ? currentIndex.value + (distance < 0 ? 1 : -1) : currentIndex.value;
      currentIndex.value = targetIndex;
      moveByIndex(targetIndex);
    };
    return {
      name: ref(name),
      gridColumn: gridColumn,
      pageNum: pageNum,
      actionItems: actionItems,
      currentIndex: currentIndex,
      containerWrapper: containerWrapper,
      wrapperStyle: wrapperStyle,
      handleSelected: handleSelected,
      handleTouchstart: handleTouchstart,
      handleTouchmove: handleTouchmove,
      handleTouchend: handleTouchend
    };
  }
});

function render(_ctx, _cache, $props, $setup, $data, $options) {
  var _component_t_grid_item = resolveComponent("t-grid-item");
  var _component_t_grid = resolveComponent("t-grid");
  return openBlock(), createElementBlock("div", {
    ref: "containerWrapper",
    class: normalizeClass("".concat(_ctx.name, "__menu-wrapper"))
  }, [createElementVNode("div", {
    class: normalizeClass("".concat(_ctx.name, "__menu-slider")),
    style: normalizeStyle(_ctx.wrapperStyle),
    onTouchstart: _cache[0] || (_cache[0] = function () {
      return _ctx.handleTouchstart && _ctx.handleTouchstart.apply(_ctx, arguments);
    }),
    onTouchmove: _cache[1] || (_cache[1] = function () {
      return _ctx.handleTouchmove && _ctx.handleTouchmove.apply(_ctx, arguments);
    }),
    onTouchend: _cache[2] || (_cache[2] = function () {
      return _ctx.handleTouchend && _ctx.handleTouchend.apply(_ctx, arguments);
    })
  }, [(openBlock(true), createElementBlock(Fragment, null, renderList(_ctx.actionItems, function (Items, i) {
    return openBlock(), createElementBlock("div", {
      key: i,
      class: normalizeClass("".concat(_ctx.name, "__menu"))
    }, [createVNode(_component_t_grid, {
      column: _ctx.gridColumn
    }, {
      default: withCtx(function () {
        return [(openBlock(true), createElementBlock(Fragment, null, renderList(Items, function (item, index) {
          return openBlock(), createBlock(_component_t_grid_item, {
            key: index,
            text: item.label,
            image: item.icon,
            onClick: function onClick($event) {
              return _ctx.handleSelected(i * _ctx.count + index);
            }
          }, null, 8, ["text", "image", "onClick"]);
        }), 128))];
      }),
      _: 2
    }, 1032, ["column"])], 2);
  }), 128))], 38), _ctx.pageNum > 1 ? (openBlock(), createElementBlock("div", {
    key: 0,
    class: normalizeClass("".concat(_ctx.name, "__indicator"))
  }, [(openBlock(true), createElementBlock(Fragment, null, renderList(_ctx.pageNum, function (index) {
    var _normalizeClass2;
    return openBlock(), createElementBlock("div", {
      key: index,
      class: normalizeClass((_normalizeClass2 = {}, _defineProperty(_normalizeClass2, "".concat(_ctx.name, "__indicator-item"), true), _defineProperty(_normalizeClass2, "on", _ctx.currentIndex === index - 1), _normalizeClass2))
    }, null, 2);
  }), 128))], 2)) : createCommentVNode("", true)], 2);
}

script.render = render;

export { script as default };
