// tdesign-mobile-vue/esm/action-sheet/menu-list.js
/**
 * tdesign v0.12.1
 * (c) 2022 TDesign Group
 * @license MIT
 */

import { defineComponent, ref, openBlock, createElementBlock, normalizeClass, Fragment, renderList, renderSlot, createElementVNode, normalizeStyle, toDisplayString } from 'vue';
import config from '../config.js';

var prefix = config.prefix;
var name = "".concat(prefix, "-action-sheet");
var script = defineComponent({
  props: {
    items: {
      type: Array,
      required: true
    }
  },
  emits: ["selected"],
  setup: function setup(props, context) {
    var handleSelected = function handleSelected(index) {
      context.emit("selected", index);
    };
    return {
      name: ref(name),
      handleSelected: handleSelected
    };
  }
});

var _hoisted_1 = ["disabled", "onClick"];
function render(_ctx, _cache, $props, $setup, $data, $options) {
  return openBlock(), createElementBlock("div", {
    class: normalizeClass("".concat(_ctx.name, "__menu"))
  }, [(openBlock(true), createElementBlock(Fragment, null, renderList(_ctx.items, function (item, index) {
    return openBlock(), createElementBlock("button", {
      key: index,
      class: normalizeClass("".concat(_ctx.name, "__cell")),
      disabled: item.disabled,
      onClick: function onClick($event) {
        return _ctx.handleSelected(index);
      }
    }, [renderSlot(_ctx.$slots, "cell", {
      item: item
    }, function () {
      return [createElementVNode("div", {
        class: normalizeClass("".concat(_ctx.name, "__cell-text")),
        style: normalizeStyle({
          color: item.color
        })
      }, toDisplayString(item.label), 7)];
    })], 10, _hoisted_1);
  }), 128))], 2);
}

script.render = render;

export { script as default };
