// tdesign-mobile-vue/esm/cascader/steps.js
/**
 * tdesign v0.12.1
 * (c) 2022 TDesign Group
 * @license MIT
 */

import { defineComponent, openBlock, createElementBlock, Fragment, renderList, createElementVNode, normalizeClass, toDisplayString, createCommentVNode } from 'vue';

var _hoisted_1 = {
  key: 0,
  class: "t-cascader__steps"
};
var _hoisted_2 = ["onClick"];
var script = /* @__PURE__ */defineComponent({
  __name: "steps",
  props: {
    items: Array,
    value: Number
  },
  emits: ["update:value"],
  setup: function setup(__props, _ref) {
    var emit = _ref.emit;
    var handleClick = function handleClick(index) {
      emit("update:value", index);
    };
    return function (_ctx, _cache) {
      return Array.isArray(__props.items) ? (openBlock(), createElementBlock("div", _hoisted_1, [(openBlock(true), createElementBlock(Fragment, null, renderList(__props.items, function (item, index) {
        return openBlock(), createElementBlock("div", {
          key: index,
          class: "t-cascader__step",
          onClick: function onClick($event) {
            return handleClick(index);
          }
        }, [createElementVNode("div", {
          class: normalizeClass({
            "t-cascader__step-dot": true,
            "t-cascader__step-dot--active": index === __props.value,
            "t-cascader__step-dot--last": index === __props.items.length - 1
          })
        }, null, 2), createElementVNode("div", {
          class: normalizeClass({
            "t-cascader__step-label": true,
            "t-cascader__step-label--active": index === __props.value
          })
        }, toDisplayString(item), 3)], 8, _hoisted_2);
      }), 128))])) : createCommentVNode("", true);
    };
  }
});

export { script as default };
