// tdesign-mobile-vue/lib/popup/props.js
/**
 * tdesign v0.12.1
 * (c) 2022 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var popupProps = {
  closeOnOverlayClick: {
    type: Boolean,
    default: true
  },
  customStyle: {
    type: String,
    default: ""
  },
  lockScroll: {
    type: Boolean,
    default: true
  },
  overlayProps: {
    type: Object,
    default: function _default() {
      return {};
    }
  },
  placement: {
    type: String,
    default: "top",
    validator: function validator(val) {
      if (!val) return true;
      return ["top", "left", "right", "bottom", "center"].includes(val);
    }
  },
  showOverlay: {
    type: Boolean,
    default: true
  },
  to: String,
  transitionName: {
    type: String,
    default: ""
  },
  visible: {
    type: Boolean,
    default: void 0
  },
  modelValue: {
    type: Boolean,
    default: void 0
  },
  defaultVisible: Boolean,
  zIndex: {
    type: Number
  },
  onClose: Function,
  onClosed: Function,
  onOpen: Function,
  onOpened: Function,
  onVisibleChange: Function
};

exports["default"] = popupProps;
