// tdesign-mobile-vue/lib/button/props.js
/**
 * tdesign v0.12.1
 * (c) 2022 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var ButtonProps = {
  block: Boolean,
  content: {
    type: [String, Function]
  },
  disabled: Boolean,
  ghost: Boolean,
  icon: {
    type: Function
  },
  loading: Boolean,
  shape: {
    type: String,
    default: "rectangle",
    validator: function validator(val) {
      if (!val) return true;
      return ["rectangle", "square", "round", "circle"].includes(val);
    }
  },
  size: {
    type: String,
    default: "medium",
    validator: function validator(val) {
      if (!val) return true;
      return ["small", "medium", "large"].includes(val);
    }
  },
  theme: {
    type: String,
    default: "default",
    validator: function validator(val) {
      if (!val) return true;
      return ["default", "primary", "danger"].includes(val);
    }
  },
  variant: {
    type: String,
    default: "base",
    validator: function validator(val) {
      if (!val) return true;
      return ["base", "outline", "text"].includes(val);
    }
  },
  onClick: Function
};

exports["default"] = ButtonProps;
