// tdesign-mobile-vue/es/message/index.js
/**
 * tdesign v0.12.1
 * (c) 2022 TDesign Group
 * @license MIT
 */

import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { ref, defineComponent, h, createApp, nextTick } from 'vue';
import script from './message.js';
import './style/css.js';
import '@babel/runtime/helpers/slicedToArray';
import 'tdesign-icons-vue-next';
import './props.js';
import '../config.js';
import '../shared/index.js';
import '../shared/functions.js';
import '../shared/util.js';
import '../shared/component.js';
import '../shared/constants.js';
import '../shared/render.js';
import '@babel/runtime/helpers/typeof';
import 'lodash/camelCase';
import '../shared/render-tnode.js';
import '../shared/useToggle/index.js';
import '../shared/useCountDown/index.js';
import '@babel/runtime/helpers/asyncToGenerator';
import '@babel/runtime/regenerator';
import '@vueuse/core';
import '../shared/useCountDown/utils.js';
import '../shared/useDefault/index.js';
import '../shared/useEmitEvent/index.js';
import '../shared/useChildSlots/index.js';
import '@babel/runtime/helpers/toConsumableArray';
import '../shared/useVModel/index.js';
import '../shared/useTouch/index.js';
import '../shared/useScrollParent/index.js';
import '../shared/useExpose/index.js';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function create(props) {
  var visible = ref(false);
  var root = document.createElement("div");
  document.body.appendChild(root);
  var component = defineComponent({
    render: function render() {
      return h(script, _objectSpread(_objectSpread({}, props), {}, {
        visible: visible.value,
        onClose: function onClose() {
          visible.value = false;
        },
        onClosed: function onClosed() {
          root.remove();
        }
      }));
    }
  });
  createApp(component).mount(root);
  nextTick(function () {
    visible.value = true;
  });
}
var defaultProps = {
  align: "left",
  closeBtn: "",
  content: "",
  duration: 3e3,
  theme: "info",
  visible: false,
  zIndex: 5e3,
  onClose: function onClose() {},
  onClosed: function onClosed() {},
  onOpen: function onOpen() {},
  onOpened: function onOpened() {},
  onChange: function onChange() {}
};
["info", "success", "warning", "error"].forEach(function (theme) {
  script[theme] = function (options) {
    var props = _objectSpread(_objectSpread({}, defaultProps), {}, {
      theme: theme
    });
    if (typeof options === "string") {
      props.content = options;
    } else {
      props = _objectSpread(_objectSpread({}, props), options);
    }
    create(props);
  };
});
script.install = function (app) {
  var name = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : "";
  app.component(name || script.name, script);
  app.config.globalProperties.$message = script;
};
var MessagePlugin = script;

export { MessagePlugin, MessagePlugin as default };
