// tdesign-mobile-vue/lib/pull-down-refresh/__test__/utils.js
/**
 * tdesign v0.12.1
 * (c) 2022 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _classCallCheck = require('@babel/runtime/helpers/classCallCheck');
var _createClass = require('@babel/runtime/helpers/createClass');
var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var vue = require('vue');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _classCallCheck__default = /*#__PURE__*/_interopDefaultLegacy(_classCallCheck);
var _createClass__default = /*#__PURE__*/_interopDefaultLegacy(_createClass);
var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

function getTouch(el, x, y) {
  return {
    identifier: Date.now(),
    target: el,
    pageX: x,
    pageY: y,
    clientX: x,
    clientY: y,
    radiusX: 2.5,
    radiusY: 2.5,
    rotationAngle: 10,
    force: 0.5
  };
}
function trigger(wrapper, eventName) {
  var x = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 0;
  var y = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : 0;
  var options = arguments.length > 4 && arguments[4] !== undefined ? arguments[4] : {};
  var el = "element" in wrapper ? wrapper.element : wrapper;
  var touchList = options.touchList || [getTouch(el, x, y)];
  if (options.x || options.y) {
    touchList.push(getTouch(el, options.x, options.y));
  }
  var event = document.createEvent("CustomEvent");
  event.initCustomEvent(eventName, true, true, {});
  Object.assign(event, {
    clientX: x,
    clientY: y,
    touches: touchList,
    targetTouches: touchList,
    changedTouches: touchList
  });
  el.dispatchEvent(event);
  return vue.nextTick();
}
var MockResizeObserver = /*#__PURE__*/function () {
  function MockResizeObserver(callback) {
    var _this = this;
    _classCallCheck__default["default"](this, MockResizeObserver);
    _defineProperty__default["default"](this, "_callback", void 0);
    _defineProperty__default["default"](this, "_element", void 0);
    _defineProperty__default["default"](this, "trigger", function (event) {
      var _event$width = event.width,
        width = _event$width === void 0 ? 0 : _event$width,
        _event$height = event.height,
        height = _event$height === void 0 ? 0 : _event$height;
      _this._callback([{
        contentRect: {
          width: width,
          height: height
        }
      }]);
    });
    this._callback = callback;
  }
  _createClass__default["default"](MockResizeObserver, [{
    key: "observe",
    value: function observe(element) {
      this._element = element;
      this._element.addEventListener("resize", this.trigger);
    }
  }, {
    key: "unobserve",
    value: function unobserve() {
      this._element.removeEventListener("resize", this.trigger);
    }
  }, {
    key: "disconnect",
    value: function disconnect() {
      this._element.removeEventListener("resize", this.trigger);
    }
  }]);
  return MockResizeObserver;
}();

exports.MockResizeObserver = MockResizeObserver;
exports.trigger = trigger;
