// tdesign-mobile-vue/lib/toast/index.js
/**
 * tdesign v0.12.1
 * (c) 2022 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var vue = require('vue');
var toast_toast = require('./toast.js');
require('tdesign-icons-vue-next');
require('../overlay/index.js');
require('../overlay/overlay.js');
require('../config.js');
require('../overlay/props.js');
require('../shared/component.js');
require('./props.js');
require('../shared/render-tnode.js');
require('../shared/render.js');
require('@babel/runtime/helpers/typeof');
require('lodash/camelCase');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty__default["default"](target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var instance = null;
var app;
function Toast(props) {
  var root = document.createElement("div");
  document.body.appendChild(root);
  var propsObject = _objectSpread({
    duration: 2e3
  }, parseOptions(props));
  if (instance) {
    instance.clear();
  }
  instance = toast_toast["default"];
  instance.clear = function () {
    clearTimeout(instance.timer);
    app.unmount();
    root.remove();
  };
  if (propsObject.duration && propsObject.duration > 0) {
    instance.timer = setTimeout(instance.clear, propsObject.duration);
  }
  app = vue.createApp(instance, _objectSpread({}, propsObject));
  app.mount(root);
  return instance;
}
Toast.clear = function () {
  if (instance) {
    instance.clear();
  }
};
["loading", "success", "fail"].forEach(function (type) {
  if (!type) {
    return;
  }
  Toast[type] = function (options) {
    var props = {
      message: "",
      theme: type
    };
    if (typeof options === "string") {
      props.message = options;
    } else {
      props = _objectSpread(_objectSpread({}, props), options);
    }
    return Toast(props);
  };
});
function parseOptions(message) {
  if (typeof message === "string") {
    return {
      message: message
    };
  }
  return message;
}
Toast.install = function (app2) {
  app2.config.globalProperties.$toast = Toast;
};
var ToastPlugin = Toast;

exports.ToastPlugin = ToastPlugin;
exports["default"] = ToastPlugin;
