// tdesign-mobile-vue/lib/toast/props.js
/**
 * tdesign v0.12.1
 * (c) 2022 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var ToastProps = {
  direction: {
    type: String,
    default: "row",
    validator: function validator(val) {
      if (!val) return true;
      return ["row", "column"].includes(val);
    }
  },
  duration: {
    type: Number,
    default: 2e3
  },
  icon: {
    type: [String, Function]
  },
  message: {
    type: [String, Function]
  },
  overlayProps: {
    type: Object,
    default: function _default() {
      return {};
    }
  },
  placement: {
    type: String,
    default: "middle",
    validator: function validator(val) {
      if (!val) return true;
      return ["top", "middle", "bottom"].includes(val);
    }
  },
  preventScrollThrough: Boolean,
  showOverlay: Boolean,
  theme: {
    type: String,
    validator: function validator(val) {
      if (!val) return true;
      return ["loading", "success", "fail"].includes(val);
    }
  }
};

exports["default"] = ToastProps;
