// tdesign-mobile-vue/es/input/props.js
/**
 * tdesign v0.12.1
 * (c) 2022 TDesign Group
 * @license MIT
 */

var InputProps = {
  align: {
    type: String,
    default: "left",
    validator: function validator(val) {
      return ["left", "center", "right"].includes(val);
    }
  },
  autocomplete: Boolean,
  autofocus: Boolean,
  clearable: Boolean,
  disabled: Boolean,
  errorMessage: {
    type: String,
    default: ""
  },
  label: {
    type: [String, Function]
  },
  maxcharacter: {
    type: Number
  },
  maxlength: {
    type: Number
  },
  name: {
    type: String,
    default: ""
  },
  placeholder: {
    type: String,
    default: void 0
  },
  prefixIcon: {
    type: Function
  },
  readonly: Boolean,
  required: Boolean,
  size: {
    type: String,
    default: "small",
    validator: function validator(val) {
      return ["small", "medium"].includes(val);
    }
  },
  suffix: {
    type: [String, Function]
  },
  suffixIcon: {
    type: Function
  },
  type: {
    type: String,
    default: "text",
    validator: function validator(val) {
      return ["text", "number", "url", "tel", "password", "search", "submit", "hidden"].includes(val);
    }
  },
  value: {
    type: [String, Number]
  },
  modelValue: {
    type: [String, Number]
  },
  defaultValue: {
    type: [String, Number]
  },
  onBlur: Function,
  onChange: Function,
  onClear: Function,
  onFocus: Function
};

export { InputProps as default };
