// tdesign-mobile-vue/lib/action-sheet/menu-list.js
/**
 * tdesign v0.12.1
 * (c) 2022 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var vue = require('vue');
var config = require('../config.js');

var prefix = config["default"].prefix;
var name = "".concat(prefix, "-action-sheet");
var script = vue.defineComponent({
  props: {
    items: {
      type: Array,
      required: true
    }
  },
  emits: ["selected"],
  setup: function setup(props, context) {
    var handleSelected = function handleSelected(index) {
      context.emit("selected", index);
    };
    return {
      name: vue.ref(name),
      handleSelected: handleSelected
    };
  }
});

var _hoisted_1 = ["disabled", "onClick"];
function render(_ctx, _cache, $props, $setup, $data, $options) {
  return vue.openBlock(), vue.createElementBlock("div", {
    class: vue.normalizeClass("".concat(_ctx.name, "__menu"))
  }, [(vue.openBlock(true), vue.createElementBlock(vue.Fragment, null, vue.renderList(_ctx.items, function (item, index) {
    return vue.openBlock(), vue.createElementBlock("button", {
      key: index,
      class: vue.normalizeClass("".concat(_ctx.name, "__cell")),
      disabled: item.disabled,
      onClick: function onClick($event) {
        return _ctx.handleSelected(index);
      }
    }, [vue.renderSlot(_ctx.$slots, "cell", {
      item: item
    }, function () {
      return [vue.createElementVNode("div", {
        class: vue.normalizeClass("".concat(_ctx.name, "__cell-text")),
        style: vue.normalizeStyle({
          color: item.color
        })
      }, vue.toDisplayString(item.label), 7)];
    })], 10, _hoisted_1);
  }), 128))], 2);
}

script.render = render;

exports["default"] = script;
