// tdesign-mobile-vue/lib/upload/props.js
/**
 * tdesign v0.12.1
 * (c) 2022 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var UploadProps = {
  accept: {
    type: String,
    default: ""
  },
  action: {
    type: String,
    default: ""
  },
  allowUploadDuplicateFile: Boolean,
  autoUpload: {
    type: Boolean,
    default: true
  },
  beforeUpload: {
    type: Function
  },
  data: {
    type: Object
  },
  deleteBtn: {
    type: [String, Function]
  },
  disabled: Boolean,
  fileListDisplay: {
    type: Function
  },
  files: {
    type: Array
  },
  defaultFiles: {
    type: Array
  },
  format: {
    type: Function
  },
  formatResponse: {
    type: Function
  },
  gridConfig: {
    type: Object
  },
  headers: {
    type: Object
  },
  imageProps: {
    type: Object
  },
  max: {
    type: Number,
    default: 0
  },
  method: {
    type: String,
    default: "POST",
    validator: function validator(val) {
      if (!val) return true;
      return ["POST", "GET", "PUT", "OPTION", "PATCH", "post", "get", "put", "option", "patch"].includes(val);
    }
  },
  multiple: Boolean,
  requestMethod: {
    type: Function
  },
  sizeLimit: {
    type: [Number, Object]
  },
  trigger: {
    type: [String, Function]
  },
  useMockProgress: {
    type: Boolean,
    default: true
  },
  withCredentials: Boolean,
  onCancelUpload: Function,
  onChange: Function,
  onFail: Function,
  onPreview: Function,
  onProgress: Function,
  onRemove: Function,
  onSelectChange: Function,
  onSuccess: Function
};

exports["default"] = UploadProps;
