// tdesign-mobile-vue/lib/dialog/index.js
/**
 * tdesign v0.12.1
 * (c) 2022 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var vue = require('vue');
var dialog_dialog = require('./dialog.js');
require('../button/index.js');
require('../button/button.js');
require('tdesign-icons-vue-next');
require('../shared/constants.js');
require('../config.js');
require('../button/props.js');
require('../shared/render-tnode.js');
require('../shared/useEmitEvent/index.js');
require('lodash/camelCase');
require('../shared/render.js');
require('@babel/runtime/helpers/typeof');
require('../shared/component.js');
require('../popup/index.js');
require('../popup/popup.js');
require('@babel/runtime/helpers/slicedToArray');
require('../popup/props.js');
require('../overlay/index.js');
require('../overlay/overlay.js');
require('../overlay/props.js');
require('../shared/useDefault/index.js');
require('./props.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty__default["default"](target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var DialogPropsDefault = {
  title: "\u6E29\u99A8\u63D0\u9192",
  content: "",
  type: "",
  confirmBtn: "\u786E\u8BA4",
  cancelBtn: "\u53D6\u6D88",
  visible: false,
  zIndex: 2500,
  showOverlay: true,
  width: "320px",
  closeOnOverlayClick: false
};
var instance;
function create(props) {
  var visible = vue.ref(false);
  var root = document.createElement("div");
  document.body.appendChild(root);
  var propsObject = _objectSpread(_objectSpread({}, DialogPropsDefault), typeof props === "string" ? {
    content: props
  } : props);
  if (instance) {
    instance.clear();
  }
  instance = vue.defineComponent({
    render: function render() {
      return vue.h(dialog_dialog["default"], _objectSpread(_objectSpread({}, propsObject), {}, {
        visible: visible.value,
        onConfirm: function onConfirm(context) {
          if (typeof propsObject.onConfirm === "function") {
            propsObject.onConfirm(context);
          }
          visible.value = false;
        },
        onCancel: function onCancel(context) {
          if (typeof propsObject.onCancel === "function") {
            propsObject.onCancel(context);
          }
          visible.value = false;
        },
        onOverlayClick: function onOverlayClick(context) {
          if (typeof propsObject.onOverlayClick === "function") {
            propsObject.onOverlayClick(context);
          }
          visible.value = false;
        },
        onClose: function onClose(context) {
          root.remove();
          if (typeof propsObject.onClose === "function") {
            propsObject.onClose(context);
          }
        }
      }));
    }
  });
  instance.clear = function () {
    root.remove();
  };
  vue.createApp(instance).mount(root);
  vue.nextTick(function () {
    visible.value = true;
  });
  return instance;
}
["show", "alert", "confirm"].forEach(function (type) {
  dialog_dialog["default"][type] = function (options) {
    var props = {
      content: "",
      type: type
    };
    if (typeof options === "string") {
      props.content = options;
    } else {
      props = _objectSpread(_objectSpread({}, props), options);
    }
    return create(props);
  };
});
dialog_dialog["default"].install = function (app) {
  var name = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : "";
  app.component(name || dialog_dialog["default"].name, dialog_dialog["default"]);
  app.config.globalProperties.$dialog = dialog_dialog["default"];
  app.provide("$dialog", dialog_dialog["default"]);
};
var DialogPlugin = dialog_dialog["default"];

exports.DialogPlugin = DialogPlugin;
exports.DialogPropsDefault = DialogPropsDefault;
exports["default"] = DialogPlugin;
