// tdesign-mobile-vue/lib/dialog/props.js
/**
 * tdesign v0.12.1
 * (c) 2022 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var DialogProps = {
  actions: {
    type: [Array, Function]
  },
  buttonLayout: {
    type: String,
    default: "horizontal",
    validator: function validator(val) {
      if (!val) return true;
      return ["horizontal", "vertical"].includes(val);
    }
  },
  cancelBtn: {
    type: [String, Object, Function],
    default: ""
  },
  closeOnOverlayClick: {
    type: Boolean,
    default: true
  },
  confirmBtn: {
    type: [String, Object, Function],
    default: ""
  },
  content: {
    type: [String, Function]
  },
  destroyOnClose: Boolean,
  preventScrollThrough: {
    type: Boolean,
    default: true
  },
  showInAttachedElement: Boolean,
  showOverlay: {
    type: Boolean,
    default: true
  },
  title: {
    type: [String, Function]
  },
  visible: Boolean,
  width: {
    type: [String, Number]
  },
  zIndex: {
    type: Number
  },
  onCancel: Function,
  onClose: Function,
  onConfirm: Function,
  onOverlayClick: Function
};

exports["default"] = DialogProps;
