// tdesign-mobile-vue/lib/swiper/props.js
/**
 * tdesign v0.12.1
 * (c) 2022 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var SwiperProps = {
  animation: {
    type: String,
    default: "slide",
    validator: function validator(val) {
      if (!val) return true;
      return ["slide"].includes(val);
    }
  },
  autoplay: {
    type: Boolean,
    default: true
  },
  current: {
    type: Number,
    default: void 0
  },
  modelValue: {
    type: Number,
    default: void 0
  },
  defaultCurrent: {
    type: Number,
    default: 0
  },
  direction: {
    type: String,
    default: "horizontal",
    validator: function validator(val) {
      if (!val) return true;
      return ["horizontal", "vertical"].includes(val);
    }
  },
  duration: {
    type: Number,
    default: 300
  },
  height: {
    type: Number
  },
  interval: {
    type: Number,
    default: 5e3
  },
  loop: {
    type: Boolean,
    default: true
  },
  navigation: {
    type: [Object, Function]
  },
  paginationPosition: {
    type: String,
    default: "bottom",
    validator: function validator(val) {
      if (!val) return true;
      return ["top-left", "top", "top-right", "bottom-left", "bottom", "bottom-right"].includes(val);
    }
  },
  onChange: Function
};

exports["default"] = SwiperProps;
