// tdesign-mobile-vue/lib/date-time-picker/date-time-picker.js
/**
 * tdesign v0.12.1
 * (c) 2022 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _toConsumableArray = require('@babel/runtime/helpers/toConsumableArray');
var _typeof = require('@babel/runtime/helpers/typeof');
var _slicedToArray = require('@babel/runtime/helpers/slicedToArray');
var vue = require('vue');
var dayjs = require('dayjs');
var customParseFormat = require('../_chunks/dep-f6dcad97.js');
var config = require('../config.js');
var dateTimePicker_props = require('./props.js');
var picker_index = require('../picker/index.js');
var shared_useEmitEvent_index = require('../shared/useEmitEvent/index.js');
var shared_useVModel_index = require('../shared/useVModel/index.js');
require('../picker/picker.js');
require('../picker/props.js');
require('../button/index.js');
require('../button/button.js');
require('@babel/runtime/helpers/defineProperty');
require('tdesign-icons-vue-next');
require('../shared/constants.js');
require('../button/props.js');
require('../shared/render-tnode.js');
require('../shared/render.js');
require('lodash/camelCase');
require('../shared/component.js');
require('../picker/picker-item.js');
require('../picker/picker.class.js');
require('@babel/runtime/helpers/classCallCheck');
require('@babel/runtime/helpers/createClass');
require('../shared/useExpose/index.js');
require('../shared/useChildSlots/index.js');
require('../picker/cascade.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _toConsumableArray__default = /*#__PURE__*/_interopDefaultLegacy(_toConsumableArray);
var _typeof__default = /*#__PURE__*/_interopDefaultLegacy(_typeof);
var _slicedToArray__default = /*#__PURE__*/_interopDefaultLegacy(_slicedToArray);
var dayjs__default = /*#__PURE__*/_interopDefaultLegacy(dayjs);

var weekday$1 = {exports: {}};

(function (module, exports) {
  !function (e, t) {
    module.exports = t() ;
  }(customParseFormat.commonjsGlobal, function () {

    return function (e, t) {
      t.prototype.weekday = function (e) {
        var t = this.$locale().weekStart || 0,
          i = this.$W,
          n = (i < t ? i + 7 : i) - t;
        return this.$utils().u(e) ? n : this.subtract(n, "day").add(e, "day");
      };
    };
  });
})(weekday$1);
var weekday = weekday$1.exports;

dayjs__default["default"].extend(weekday);
dayjs__default["default"].extend(customParseFormat.customParseFormat);
var prefix = config["default"].prefix;
var name = "".concat(prefix, "-date-time-picker");
var precisionRankRecord = {
  year: 0,
  month: 1,
  date: 2,
  hour: 3,
  minute: 4,
  second: 5
};
var script = vue.defineComponent({
  name: name,
  components: {
    TPicker: picker_index.Picker
  },
  props: dateTimePicker_props["default"],
  emits: ["change", "cancel", "confirm", "pick", "update:modelValue", "update:value"],
  setup: function setup(props, context) {
    var emitEvent = shared_useEmitEvent_index.useEmitEvent(props, context.emit);
    var pickeInstance = vue.ref(null);
    var isChanged = vue.ref(false);
    var realDateValue = vue.ref();
    var className = vue.computed(function () {
      return ["".concat(name)];
    });
    var _toRefs = vue.toRefs(props),
      value = _toRefs.value,
      modelValue = _toRefs.modelValue;
    var _useVModel = shared_useVModel_index.useVModel(value, modelValue, props.defaultValue, props.onChange),
      _useVModel2 = _slicedToArray__default["default"](_useVModel, 2),
      dateTimePickerValue = _useVModel2[0],
      setDateTimePickerValue = _useVModel2[1];
    var title = vue.computed(function () {
      return props.title || "\u9009\u62E9\u65F6\u95F4";
    });
    var confirmButtonText = vue.computed(function () {
      return props.confirmBtn || "\u786E\u5B9A";
    });
    var cancelButtonText = vue.computed(function () {
      return props.cancelBtn || "\u53D6\u6D88";
    });
    var start = vue.computed(function () {
      return props.start ? dayjs__default["default"](props.start) : dayjs__default["default"]().subtract(10, "year");
    });
    var end = vue.computed(function () {
      return props.end ? dayjs__default["default"](props.end) : dayjs__default["default"]().add(10, "year");
    });
    var renderLabel = vue.computed(function () {
      return props.renderLabel;
    });
    var isPrecision = function isPrecision(type) {
      if (!props.mode) {
        return false;
      }
      switch (type) {
        case "year":
          return typeof props.mode === "string" && precisionRankRecord[props.mode] >= 0 || _typeof__default["default"](props.mode) === "object" && precisionRankRecord[props.mode[0]] >= 0;
        case "month":
          return typeof props.mode === "string" && precisionRankRecord[props.mode] >= 1 || _typeof__default["default"](props.mode) === "object" && precisionRankRecord[props.mode[0]] >= 1;
        case "date":
          return typeof props.mode === "string" && precisionRankRecord[props.mode] >= 2 || _typeof__default["default"](props.mode) === "object" && precisionRankRecord[props.mode[0]] >= 2;
        case "hour":
          return typeof props.mode === "string" && precisionRankRecord[props.mode] >= 3 || _typeof__default["default"](props.mode) === "object" && precisionRankRecord[props.mode[1]] >= 3;
        case "minute":
          return typeof props.mode === "string" && precisionRankRecord[props.mode] >= 4 || _typeof__default["default"](props.mode) === "object" && precisionRankRecord[props.mode[1]] >= 4;
        case "second":
          return typeof props.mode === "string" && precisionRankRecord[props.mode] >= 5 || _typeof__default["default"](props.mode) === "object" && precisionRankRecord[props.mode[1]] >= 5;
        default:
          return true;
      }
    };
    var getPickerValueByDateTimePickerValue = function getPickerValueByDateTimePickerValue(value2) {
      var currentDate = dayjs__default["default"](value2);
      var ret = [];
      Object.keys(precisionRankRecord).forEach(function (item) {
        if (isPrecision(item)) {
          ret.push("".concat(currentDate[item]()));
        }
      });
      return ret;
    };
    var pickerValue = vue.ref(getPickerValueByDateTimePickerValue(dateTimePickerValue.value || start.value.valueOf()));
    var lastTimePicker = _toConsumableArray__default["default"](pickerValue.value);
    var currentPicker = _toConsumableArray__default["default"](pickerValue.value);
    var getDateTimePickerValueByPickerValue = function getDateTimePickerValueByPickerValue(value2) {
      var valueLength = 0;
      var date = dayjs__default["default"]();
      Object.keys(precisionRankRecord).forEach(function (item, index) {
        if (isPrecision(item)) {
          date = date[item](value2[valueLength]);
          valueLength++;
        }
      });
      return date;
    };
    var generateDatePickerColumns = function generateDatePickerColumns(selected, min, max, renderLabel2) {
      var ret = [];
      var minYear = min.year();
      var minMonth = min.month() + 1;
      var minDay = min.date();
      var minHour = min.hour();
      var minMinute = min.minute();
      var minSecond = min.second();
      var maxYear = max.year();
      var maxMonth = max.month() + 1;
      var maxDay = max.date();
      var maxHour = max.hour();
      var maxMinute = max.minute();
      var maxSecond = max.second();
      var selectedDate = {};
      var selectedLength = 0;
      Object.keys(precisionRankRecord).forEach(function (item) {
        var newKey = "selected".concat(item.substr(0, 1).toUpperCase()).concat(item.substr(1, item.length));
        if (isPrecision(item)) {
          selectedDate[newKey] = parseInt("".concat(selected[selectedLength]), 10);
          selectedLength++;
        } else {
          selectedDate[newKey] = void 0;
        }
      });
      var isInMinYear = selectedDate.selectedYear === minYear;
      var isInMaxYear = selectedDate.selectedYear === maxYear;
      var isInMinMonth = isInMinYear && selectedDate.selectedMonth + 1 === minMonth;
      var isInMaxMonth = isInMaxYear && selectedDate.selectedMonth + 1 === maxMonth;
      var isInMinDay = isInMinMonth && selectedDate.selectedDay === minDay;
      var isInMaxDay = isInMaxMonth && selectedDate.selectedDay === maxDay;
      var isInMinHour = isInMinDay && selectedDate.selectedHour === minHour;
      var isInMaxHour = isInMaxDay && selectedDate.selectedHour === maxHour;
      var isInMinMinute = isInMinHour && selectedDate.selectedMinute === minMinute;
      var isInMaxMinute = isInMaxHour && selectedDate.selectedMinute === maxMinute;
      var generateColumn = function generateColumn(start2, end2, type) {
        var arr = [];
        for (var i = start2; i <= end2; i++) {
          var value2 = i.toString();
          arr.push({
            label: renderLabel2 ? renderLabel2(type, i) : value2,
            value: type === "month" ? "".concat(+value2 - 1) : value2
          });
        }
        ret.push(arr);
      };
      if (isPrecision("year")) {
        generateColumn(minYear, maxYear, "year");
      }
      if (isPrecision("month")) {
        var lower = isInMinYear ? minMonth : 1;
        var upper = isInMaxYear ? maxMonth : 12;
        generateColumn(lower, upper, "month");
      }
      if (isPrecision("date")) {
        var _lower = isInMinMonth ? minDay : 1;
        var _upper = isInMaxMonth ? maxDay : dayjs__default["default"]("".concat(selected[0], "-").concat(+selected[1] + 1)).daysInMonth();
        generateColumn(_lower, _upper, "date");
      }
      if (isPrecision("hour")) {
        var _lower2 = isInMinDay ? minHour : 0;
        var _upper2 = isInMaxDay ? maxHour : 23;
        generateColumn(_lower2, _upper2, "hour");
      }
      if (isPrecision("minute")) {
        var _lower3 = isInMinHour ? minMinute : 0;
        var _upper3 = isInMaxHour ? maxMinute : 59;
        generateColumn(_lower3, _upper3, "minute");
      }
      if (isPrecision("second")) {
        var _lower4 = isInMinMinute ? minSecond : 0;
        var _upper4 = isInMaxMinute ? maxSecond : 59;
        generateColumn(_lower4, _upper4, "second");
      }
      return ret;
    };
    var onConfirm = function onConfirm(value2, context2) {
      lastTimePicker = _toConsumableArray__default["default"](currentPicker);
      var currentDate = getDateTimePickerValueByPickerValue(value2);
      emitEvent("confirm", dayjs__default["default"](currentDate).format(props.format));
    };
    var onCancel = function onCancel(context2) {
      currentPicker = _toConsumableArray__default["default"](lastTimePicker);
      emitEvent("cancel", {
        e: context2.e
      });
    };
    var onChange = function onChange(value2, context2) {
      lastTimePicker = _toConsumableArray__default["default"](currentPicker);
      var currentDate = getDateTimePickerValueByPickerValue(value2);
      realDateValue.value = dayjs__default["default"](currentDate).format(props.format);
      isChanged.value = true;
    };
    var onPick = function onPick(value2, context2) {
      currentPicker = value2;
      var currentDate = getDateTimePickerValueByPickerValue(value2);
      emitEvent("pick", dayjs__default["default"](currentDate).format(props.format));
    };
    vue.watch(function () {
      return dateTimePickerValue;
    }, function (val) {
      vue.nextTick(function () {
        if (isChanged.value) {
          isChanged.value = false;
        } else {
          var _pickeInstance$value;
          (_pickeInstance$value = pickeInstance.value) === null || _pickeInstance$value === void 0 ? void 0 : _pickeInstance$value.setValues(getPickerValueByDateTimePickerValue(val.value || start.value.valueOf()));
          currentPicker = _toConsumableArray__default["default"](vue.ref(getPickerValueByDateTimePickerValue(val.value || start.value.valueOf())).value);
          lastTimePicker = _toConsumableArray__default["default"](currentPicker);
          isChanged.value = false;
        }
      });
    }, {
      immediate: true,
      deep: true
    });
    vue.watch(function () {
      return isChanged;
    }, function (val) {
      if (val.value) {
        setDateTimePickerValue(realDateValue.value);
      }
    }, {
      immediate: true,
      deep: true
    });
    return {
      pickeInstance: pickeInstance,
      className: className,
      confirmButtonText: confirmButtonText,
      cancelButtonText: cancelButtonText,
      title: title,
      start: start,
      end: end,
      renderLabel: renderLabel,
      pickerValue: pickerValue,
      currentPicker: currentPicker,
      realDateValue: realDateValue,
      isChanged: isChanged,
      generateDatePickerColumns: generateDatePickerColumns,
      onConfirm: onConfirm,
      onCancel: onCancel,
      onPick: onPick,
      onChange: onChange
    };
  }
});

function render(_ctx, _cache, $props, $setup, $data, $options) {
  var _component_t_picker = vue.resolveComponent("t-picker");
  return vue.openBlock(), vue.createElementBlock("div", {
    class: vue.normalizeClass(_ctx.className)
  }, [vue.createVNode(_component_t_picker, {
    ref: "pickeInstance",
    value: _ctx.currentPicker,
    title: _ctx.title,
    columns: function columns(selected) {
      return _ctx.generateDatePickerColumns(selected, _ctx.start, _ctx.end, _ctx.renderLabel);
    },
    onChange: _ctx.onChange,
    onConfirm: _ctx.onConfirm,
    onCancel: _ctx.onCancel,
    onPick: _ctx.onPick
  }, null, 8, ["value", "title", "columns", "onChange", "onConfirm", "onCancel", "onPick"])], 2);
}

script.render = render;

exports["default"] = script;
