// tdesign-mobile-vue/lib/tabs/tab-panel.js
/**
 * tdesign v0.12.1
 * (c) 2022 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var vue = require('vue');
var config = require('../config.js');
var tabs_tabPanelProps = require('./tab-panel-props.js');
var shared_renderTnode = require('../shared/render-tnode.js');
var shared_render = require('../shared/render.js');
require('@babel/runtime/helpers/typeof');
require('lodash/camelCase');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty__default["default"](target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var prefix = config["default"].prefix;
var name = "".concat(prefix, "-tab-panel");
var script = vue.defineComponent({
  name: name,
  components: {
    TNode: shared_renderTnode["default"]
  },
  props: tabs_tabPanelProps["default"],
  setup: function setup(props) {
    var currentValue = vue.inject("currentValue", "");
    var isActive = vue.computed(function () {
      return currentValue.value === props.value;
    });
    var internalInstance = vue.getCurrentInstance();
    var panelContent = vue.computed(function () {
      return shared_render.renderContent(internalInstance, "default", "panel");
    });
    var tabPanelClass = vue.computed(function () {
      return ["".concat(name), "".concat(prefix, "-tabs__panel")];
    });
    return _objectSpread({
      prefix: prefix,
      name: name,
      isActive: isActive,
      panelContent: panelContent,
      tabPanelClass: tabPanelClass
    }, vue.toRefs(props));
  }
});

var _hoisted_1 = ["value"];
function render(_ctx, _cache, $props, $setup, $data, $options) {
  var _component_t_node = vue.resolveComponent("t-node");
  return _ctx.destroyOnHide || _ctx.isActive ? vue.withDirectives((vue.openBlock(), vue.createElementBlock("div", {
    key: 0,
    class: vue.normalizeClass(_ctx.tabPanelClass),
    value: _ctx.value
  }, [vue.renderSlot(_ctx.$slots, "default", {}, function () {
    return [vue.createVNode(_component_t_node, {
      content: _ctx.panelContent
    }, null, 8, ["content"])];
  })], 10, _hoisted_1)), [[vue.vShow, _ctx.isActive]]) : vue.createCommentVNode("", true);
}

script.render = render;

exports["default"] = script;
