// tdesign-mobile-vue/lib/skeleton/skeleton.js
/**
 * tdesign v0.12.1
 * (c) 2022 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var _toConsumableArray = require('@babel/runtime/helpers/toConsumableArray');
var vue = require('vue');
var config = require('../config.js');
var skeleton_props = require('./props.js');
var shared_renderTnode = require('../shared/render-tnode.js');
var shared_util = require('../shared/util.js');
var shared_render = require('../shared/render.js');
require('@babel/runtime/helpers/typeof');
require('lodash/camelCase');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);
var _toConsumableArray__default = /*#__PURE__*/_interopDefaultLegacy(_toConsumableArray);

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty__default["default"](target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var prefix = config["default"].prefix;
var name = "".concat(prefix, "-skeleton");
var ThemeMap = {
  avatar: [{
    type: "circle",
    height: "64px",
    width: "64px"
  }],
  image: [{
    type: "rect",
    height: "64px",
    width: "64px"
  }],
  text: [1, [{
    width: "24%",
    height: "16px",
    marginRight: "16px"
  }, {
    width: "76%",
    height: "16px"
  }]],
  paragraph: [1, 1, 1, {
    width: "55%"
  }]
};
var script = vue.defineComponent({
  name: name,
  components: {
    TNode: shared_renderTnode["default"]
  },
  props: skeleton_props["default"],
  setup: function setup(props) {
    var _toRefs = vue.toRefs(props),
      loading = _toRefs.loading,
      animation = _toRefs.animation;
    var showContent = vue.computed(function () {
      return !loading.value;
    });
    var internalInstance = vue.getCurrentInstance();
    var skeletonContent = vue.computed(function () {
      return shared_render.renderContent(internalInstance, "default", "content");
    });
    var baseClass = name;
    var rootClasses = vue.computed(function () {
      return ["".concat(name)];
    });
    var rowCols = vue.ref([]);
    vue.watchEffect(function () {
      var _props$rowCol;
      if ((_props$rowCol = props.rowCol) !== null && _props$rowCol !== void 0 && _props$rowCol.length) {
        rowCols.value = _toConsumableArray__default["default"](props.rowCol);
      } else {
        rowCols.value = _toConsumableArray__default["default"](ThemeMap[props.theme || "text"]);
      }
    });
    var getColItemClass = function getColItemClass(obj) {
      return ["".concat(name, "__col"), "".concat(name, "--type-").concat(obj.type || "text"), _defineProperty__default["default"]({}, "".concat(name, "--animation-").concat(animation.value), animation.value)];
    };
    var getColItemStyle = function getColItemStyle(obj) {
      var styleName = ["width", "height", "marginRight", "marginLeft", "margin", "size", "background", "backgroundColor", "borderRadius"];
      var style = {};
      styleName.forEach(function (name2) {
        if (name2 in obj) {
          var px = shared_util.isNumber(obj[name2]) ? "".concat(obj[name2], "px") : obj[name2];
          if (name2 === "size") {
            var _ref2 = [px, px];
            style.width = _ref2[0];
            style.height = _ref2[1];
          } else {
            style[name2] = px;
          }
        }
      });
      return style;
    };
    var parsedRowcols = vue.computed(function () {
      return rowCols.value.map(function (item) {
        if (shared_util.isNumber(item)) {
          return [{
            class: getColItemClass({
              type: "text"
            }),
            style: {}
          }];
        }
        if (Array.isArray(item)) {
          return item.map(function (col) {
            return _objectSpread(_objectSpread({}, col), {}, {
              class: getColItemClass(col),
              style: getColItemStyle(col)
            });
          });
        }
        var nItem = item;
        return [_objectSpread(_objectSpread({}, nItem), {}, {
          class: getColItemClass(nItem),
          style: getColItemStyle(nItem)
        })];
      });
    });
    return {
      baseClass: baseClass,
      rootClasses: rootClasses,
      parsedRowcols: parsedRowcols,
      showContent: showContent,
      skeletonContent: skeletonContent
    };
  }
});

function render(_ctx, _cache, $props, $setup, $data, $options) {
  var _component_t_node = vue.resolveComponent("t-node");
  return vue.openBlock(), vue.createElementBlock("div", {
    class: vue.normalizeClass(_ctx.rootClasses)
  }, [_ctx.showContent ? (vue.openBlock(), vue.createBlock(_component_t_node, {
    key: 0,
    content: _ctx.skeletonContent
  }, null, 8, ["content"])) : (vue.openBlock(), vue.createElementBlock(vue.Fragment, {
    key: 1
  }, [_ctx.parsedRowcols.length ? (vue.openBlock(), vue.createElementBlock("div", {
    key: 0,
    class: vue.normalizeClass("".concat(_ctx.baseClass, "__content"))
  }, [(vue.openBlock(true), vue.createElementBlock(vue.Fragment, null, vue.renderList(_ctx.parsedRowcols, function (row, index) {
    return vue.openBlock(), vue.createElementBlock("div", {
      key: "row-".concat(index),
      class: vue.normalizeClass("".concat(_ctx.baseClass, "__row"))
    }, [(vue.openBlock(true), vue.createElementBlock(vue.Fragment, null, vue.renderList(row, function (col, idx) {
      return vue.openBlock(), vue.createElementBlock("div", {
        key: "col-".concat(idx),
        class: vue.normalizeClass(col.class),
        style: vue.normalizeStyle(col.style)
      }, null, 6);
    }), 128))], 2);
  }), 128))], 2)) : vue.createCommentVNode("", true)], 64))], 2);
}

script.render = render;

exports["default"] = script;
