// tdesign-mobile-vue/lib/dropdown-menu/context.js
/**
 * tdesign v0.12.1
 * (c) 2022 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var DropdownMenuExpandState = /* @__PURE__ */function (DropdownMenuExpandState2) {
  DropdownMenuExpandState2["expanded"] = "expanded";
  DropdownMenuExpandState2["collapsed"] = "collapsed";
  return DropdownMenuExpandState2;
}(DropdownMenuExpandState || {});
var oldOverflow = null;
var context = {
  expandedMenuControl: null,
  recordMenuExpanded: function recordMenuExpanded(container, menuControl, action) {
    var containerDom = container;
    if (action === "expanded" /* expanded */) {
      var expandedMenuControl = this.expandedMenuControl;
      if (expandedMenuControl && expandedMenuControl !== menuControl) {
        expandedMenuControl.collapseMenu();
      }
      this.expandedMenuControl = menuControl;
      if (oldOverflow === null) {
        oldOverflow = container.style.overflow;
        containerDom.style.overflow = "hidden";
      }
    } else if (action === "collapsed" /* collapsed */) {
      if (this.expandedMenuControl === menuControl) {
        this.expandedMenuControl = null;
      }
      if (!this.expandedMenuControl && oldOverflow !== null) {
        containerDom.style.overflow = oldOverflow;
        oldOverflow = null;
      }
    }
  }
};

exports.DropdownMenuExpandState = DropdownMenuExpandState;
exports.context = context;
